scene/resources/primitive_meshes.cpp

```diff
@@ -88,6 +88,45 @@ void PrimitiveMesh::_update() const {
 		}
 	}
 
+	if (!smooth_faces) {
+		Vector<float> tangents = arr[RS::ARRAY_TANGENT];
+		Vector<Vector2> uvs = arr[RS::ARRAY_TEX_UV];
+
+		Vector<Vector3> fpoints;
+		Vector<Vector3> fnormals;
+		Vector<float> ftangents;
+		Vector<Vector2> fuvs;
+		Vector<int> findices;
+
+		const Vector3 *pr = points.ptr();
+		const float *tr = tangents.ptr();
+		const Vector2 *ur = uvs.ptr();
+		const int *ir = indices.ptr();
+
+		for (int i = 0; i < indices.size(); i += 3) {
+			Vector3 a = pr[ir[i]];
+			Vector3 b = pr[ir[i + 1]];
+			Vector3 c = pr[ir[i + 2]];
+			Vector3 normal = (a - c).cross(a - b).normalized();
+			for (int j = 0; j < 3; j++) {
+				int index = ir[i + j];
+				fpoints.push_back(pr[index]);
+				fnormals.push_back(normal);
+				for (int t = 0; t < 4; t++) {
+					ftangents.push_back(tr[index * 4 + t]);
+				}
```

Comment:
Just wondering: shouldn't tangent also be recalculated to match the new normal (face normal)? Normal+tangent+bitangent may be not orthogonal after this.